src/gluonts/dataset/jsonl.py

```diff
@@ -55,28 +56,43 @@ class JsonLinesFile:
         JSON Lines file.
     """
 
-    def __init__(self, path) -> None:
+    def __init__(self, path: Path, cache: bool = False) -> None:
         self.path = path
+        self.cache = cache
         self._len = None
-        # TODO: implement caching here
+        self._data_cache: list = []
 
     def __iter__(self):
-        with open(self.path) as jsonl_file:
-            for line_number, raw in enumerate(jsonl_file):
-                # The dataset is equally distributed among the workers
-                if not (
-                    line_number % MPWorkerInfo.num_workers
-                    == MPWorkerInfo.worker_id
-                ):
-                    continue
-
-                span = Span(path=self.path, line=line_number)
-                try:
-                    yield Line(json.loads(raw), span=span)
-                except ValueError:
-                    raise GluonTSDataError(
-                        f"Could not read json line {line_number}, {raw}"
+        # Basic idea is to split the dataset into roughly equally sized segments
+        # with lower and upper bound, where each worker is assigned one segment
+        segment_size = int(len(self) / MPWorkerInfo.num_workers)
+
+        if not self.cache or (self.cache and not self._data_cache):
+            with open(self.path) as jsonl_file:
+                for line_number, raw in enumerate(jsonl_file):
+                    lower_bound = MPWorkerInfo.worker_id * segment_size
+                    upper_bound = (
+                        (MPWorkerInfo.worker_id + 1) * segment_size
+                        if MPWorkerInfo.worker_id + 1
+                        != MPWorkerInfo.num_workers
+                        else np.inf
                     )
```

Comment:
Maybe it's better to have the exact length of the last segment be apparent here. The way `segment_size` is computed above, you could end up with the last worker having to do almost twice as much work as the other ones. For example, suppose `len(self) == 19` and `num_workers = 4`, then `segment_size = int(19 / 4) = 4`, and you'll have the first three worker process [0, 4), [4, 8) and [8, 12) respectively, and the last one [12, 19) which is 7 elements instead of 4. If all workers synchronize at some point, this may keep some workers idle for too long.

Instead, you can round up the division, and have `segment_size = int(ceiling(19 / 4)) = 5`, and have workers split the load as follows: [0, 5), [5, 10), [10, 15), [15, 19), which seems more balanced. In this case the upper bound here would become:

```python
upper_bound = min(len(self), (MPWorkerInfo.worker_id + 1) * segment_size)
```

Reply:
👍

Reply:
👍

Reply:
I like this, however this brakes the code for example for `len(dataset)=5` and `num_workers=4`, because the first two workers will get 2 each, the other two nothing, which leads to workers running into an infinite fetch loop and ultimately to a timeout.

As far as I can see the worst overhead a single worker could have is `num_workers-1` sample , which in case for large datasets is negligible, which is the main usecase for mp data leader?

What do you think?

Reply:
> As far as I can see the worst overhead a single worker could have is `num_workers-1` sample , which in case for large datasets is negligible, which is the main usecase for mp data leader?

I think the maximum overhead is actually `int(len(dataset) / num_workers) - 1` (see my example above).

Maybe to solve the case you mention, it is sufficient to change how `num_worker` is bounded by `len(dataset)`: currently we assert essentially that `num_worker <= len(dataset)` (if I'm not wrong), we could assert that `num_worker <= len(dataset)/2` or something like that

Reply:
I commited your simplified logic with `min(len(self),...` though. Turns out there was a test with a one off error too.

Reply:
`num_worker <= len(dataset)/2` should indeed work.

`int(len(dataset) / num_workers) - 1` is in fact  ` num_workers - 1` in your example, but its also the wrong formula. 

It should be impossible to get more than ` num_workers - 1` because if it was, then just distribute ` num_workers` time series among the workers, and give the rest to the last, in which case is again bounded by ` num_workers - 1`.

Or am I missing something?

Reply:
Yes, the remainder must be in {0, ..., num_workers-1}, that's true. Then I would just replace `else np.inf` with `else len(self)`, right?

Reply:
Yes!, Thx!